File: TASK07/lexer/lexer.go
package lexer

import (
	"code.skyseaee.com/sonic-self/TASK07/token"
)

type Lexer struct {
	data []byte

	offset int
	curr   rune
	pos    token.Position
	tokens chan token.Token
}

const eof = -1

func (l *Lexer) next() {
	if l.curr = l.peek(); l.curr == eof {
		return
	}

	l.offset++

	if l.curr == '\n' {
		l.pos.NextLine()
	} else {
		l.pos.Column++
	}
}

func (l *Lexer) peek() rune {
	if l.offset >= len(l.data) {
		return eof
	}

	return rune(l.data[l.offset])
}

func (l *Lexer) program() {
	for {
		var tok token.Type
		switch l.peek() {
		case eof:
			tok = token.Eof
		case '+':
			tok = token.Plus
		case '-':
			tok = token.Minus
		case '<':
			tok = token.LeftArrow
		case '>':
			tok = token.RightArrow
		case ',':
			tok = token.Comma
		case '.':
			tok = token.Period
		case '[':
			tok = token.LeftBracket
		case ']':
			tok = token.RightBracket
		default:
			l.next()
			continue
		}
		l.emit(tok)

		l.next()

		if tok == token.Eof {
			return
		}
	}
}

func (l *Lexer) emit(tok token.Type) {
	l.tokens <- token.Token{
		Type:     tok,
		Position: l.pos,
	}

	if tok == token.Eof {
		close(l.tokens)
	}
}

func Lex(data []byte) <-chan token.Token {
	l := Lexer{
		data:   data,
		pos:    token.Position{Line: 1, Column: 1},
		tokens: make(chan token.Token, 1),
	}

	go l.program()
	return l.tokens
}

File: TASK08/lexer/lexer.cpp
#pragma once
#include <iostream>
#include <vector>
#include "../token/token.cpp"
using namespace std;

static const int eof = -1;

class Lexer {
public:
    vector<char> data;
    int offset;
    char curr;
    Position pos;
    vector<Token> tokens;

    Lexer(vector<char> data): pos(0, 0) {
        offset = 0;
        curr = '.';
        this -> data = data;
    }

    char peek() {
        if(offset >= data.size()) {
            return eof;
        }

        return data[offset];
    }

    void next() {
        char curr = this->peek();
        if(curr == eof) {
            return;
        }

        offset++;
        if(curr == '\n') {
            pos.NextLine();
        } else {
            pos.column++;
        }
    }

    void emit(int token) {
        Token tok = Token(token, 1, 1);
        tokens.emplace_back(tok);
    }

    void program() {
        while(true){
            int tok;
            switch (peek())
            {
            case eof:
                tok = Eof;
            case '+':
                tok = Plus;
            case '-':
                tok = Minus;
            case '<':
                tok = LeftArrow;
            case '>':
                tok = RightArrow;
            case ',':
                tok = Comma;
            case '.':
                tok = Period;
            case '[':
                tok = LeftBracket;
            case ']':
                tok = RightBracket;
            default:
                next();
                continue;
            }
            emit(tok);
            next();

            if(tok == Eof) {
                return;
            }
        }
    }
};

